packages/expo-sqlite/src/next/SQLiteStatement.ts

```diff
@@ -0,0 +1,449 @@
+ * This is done by `Object.defineProperties` to add the properties to the AsyncGenerator.
+ */
+async function createSQLiteExecuteAsyncResult<T>(
+  database: SQLiteAnyDatabase,
+  statement: NativeStatement,
+  lastInsertRowId: number,
+  changes: number,
+  firstRowValues: SQLiteColumnValues | null
+): Promise<SQLiteExecuteAsyncResult<T>> {
+  const instance = new SQLiteExecuteAsyncResultImpl<T>(
+    database,
+    statement,
+    lastInsertRowId,
+    changes,
+    firstRowValues
+  );
+  const generator = instance.generatorAsync();
+  Object.defineProperties(generator, {
+    lastInsertRowId: { value: lastInsertRowId, enumerable: true, writable: false },
+    changes: { value: changes, enumerable: true, writable: false },
+    getFirstAsync: {
+      value: instance.getFirstAsync.bind(instance),
+      enumerable: true,
+      writable: false,
+      configurable: false,
+    },
+    getAllAsync: {
+      value: instance.getAllAsync.bind(instance),
+      enumerable: true,
+      writable: false,
+      configurable: false,
+    },
+  });
+
+  return generator as SQLiteExecuteAsyncResult<T>;
+}
+
+/**
+ * Create the `SQLiteExecuteSyncResult` instance.
+ */
+function createSQLiteExecuteSyncResult<T>(
+  database: SQLiteAnyDatabase,
+  statement: NativeStatement,
+  lastInsertRowId: number,
+  changes: number,
+  firstRowValues: SQLiteColumnValues | null
+): SQLiteExecuteSyncResult<T> {
+  const instance = new SQLiteExecuteSyncResultImpl<T>(
+    database,
+    statement,
+    lastInsertRowId,
+    changes,
+    firstRowValues
+  );
+  const generator = instance.generatorSync();
+  Object.defineProperties(generator, {
+    lastInsertRowId: { value: lastInsertRowId, enumerable: true, writable: false },
+    changes: { value: changes, enumerable: true, writable: false },
+    getFirstSync: {
+      value: instance.getFirstSync.bind(instance),
+      enumerable: true,
+      writable: false,
+      configurable: false,
+    },
+    getAllSync: {
+      value: instance.getAllSync.bind(instance),
+      enumerable: true,
+      writable: false,
+      configurable: false,
+    },
+  });
+
+  return generator as SQLiteExecuteSyncResult<T>;
+}
+
+class SQLiteExecuteAsyncResultImpl<T> {
+  private columnNames: string[] | null = null;
+
+  constructor(
+    private readonly database: SQLiteAnyDatabase,
+    private readonly statement: NativeStatement,
+    public readonly lastInsertRowId: number,
+    public readonly changes: number,
+    private firstRowValues: SQLiteColumnValues | null
+  ) {}
+
+  async getFirstAsync(): Promise<T | null> {
+    const columnNames = await this.getColumnNamesAsync();
+    const firstRowValues = this.popFirstRowValues();
+    if (firstRowValues != null) {
+      return composeRow<T>(columnNames, firstRowValues);
+    }
+    const firstRow = await this.statement.stepAsync(this.database);
+    return firstRow != null ? composeRow<T>(columnNames, firstRow) : null;
+  }
```

Comment:
It looks like this method will return each next row if called multiple times:
```ts
await statement.getFirstAsync();
await statement.getFirstAsync();
await statement.getFirstAsync();
...
```

I think this is confusing behavior -- the developer should use `db.getEachAsync()` or the statement result's async iterator if they want each row. What do you think about making `getFirstAsync()` throw if it is called more than once without resetting the statement?

Also, it would be good to document the behavior when this function is called more than once.

Similarly, what about making `getAllAsync()` throw unless the statement has been reset?

Reply:
That is a very great question! Originally, I tried to call `sqlite3_reset()` to reset the SQLite cursor and that would make getFirstAsync() as its semantic to get the "first".
However, it breaks the following test case because `sqlite3_reset()` + `sqlite3_step()` will INSERT again.
```ts
await db.execAsync('CREATE TABLE customers (id PRIMARY KEY NOT NULL, name VARCHAR(255), email VARCHAR(255));');
const statement = await db.prepareAsync('INSERT INTO customers (id, name, email) VALUES (?, ?, ?) RETURNING name, email');
const result = await statement.executeAsync(1, 'John Doe', 'john@example.com');

// ❌ violate "id" unique key if we internally call `sqlite3_reset()` + `sqlite3_step()` to get the first row, as we calling `sqlite3_step()` twice.
const row = await result.getFirstAsync();
```

The `getFirstAsync()` is actually `getNextAsync()` and `getAllAsync()` is actually `getRemainingAsync()` in our case 🤦‍♂️
Throwing errors when we cannot get the cursor moved and we cannot get the true "first" or "all" sounds good to me. I will follow up and document the throwing error cases.

Reply:
will throw errors from getFirstAsync() / getFirstSync() / getAllAsync() / getAllSync() like this: https://github.com/expo/expo/blob/a70d6107514e55bdced2e6dbcb7ef5856997457b/packages/expo-sqlite/src/next/SQLiteStatement.ts#L377-L382

since we showing that to reset the cursor in the message, i also move the resetAsync() and resetSync() to result. it makes more sense to put all the cursor related functions together.

from `await statement.resetAsync()` to
```
const result = await statement.executeAsync(...);
await result.resetAsync();
```